### PR TITLE
Fix FTBFS on hurd-i386

### PR DESCRIPTION
**Pull request by jspricke**

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Please add some description to this ticket about the rational.

**Comment by jspricke** [comment]
Quoting https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=822757

> Currently ros-rospack FTBFS on GNU/Hurd due to usage of PATH_MAX, which is not
> defined. The attached patch fixes this for all UNIX cases by dynamically
> allocating and de-allocating the needed strings. The patch works fine for UNIX,
> but has not been tested if the WIN32 or MINGW32 versions are affected. In case
> of any doubts about the patch, please forward this patch to the upstream
> authors, so I can communicate directly with them.

**Comment by dirk-thomas** [comment]
Would it be sufficient to include `linux/limits.h` which defines `PATH_MAX`?

**Comment by jspricke** [comment]
That's Linux specific.

**Comment by dirk-thomas** [comment]
The file is available via [linux-libc-dev](https://packages.debian.org/search?keywords=linux-libc-dev). Can that not just be installed?

**Comment by jspricke** [comment]
We are talking about other Unise like GNU/Hurd here, so no glibc (no entry for hurd-i386, on that site ;) )

File: src/rospack.cpp [hunk 1]
```diff
@@ -1994,23 +1994,33 @@ Rosstackage::writeCache()
   }
   else
   {
-    char tmp_cache_dir[PATH_MAX];
-    char tmp_cache_path[PATH_MAX];
-    strncpy(tmp_cache_dir, cache_path.c_str(), sizeof(tmp_cache_dir));
+    int len = cache_path.size() + 1;
```
Comment (dirk-thomas) [inline src/rospack.cpp]: This type should be `size_t`.

File: src/rospack.cpp [hunk 2]
```diff
@@ -1994,23 +1994,33 @@ Rosstackage::writeCache()
...
+    char *tmp_cache_dir = (char *)malloc(len);
```
Comment (dirk-thomas) [inline src/rospack.cpp]: Please use a C++ cast.

Same below.

File: src/rospack.cpp [hunk 3]
```diff
@@ -1994,23 +1994,33 @@ Rosstackage::writeCache()
...
+    strncpy(tmp_cache_dir, cache_path.c_str(), len);
+    // make sure tmp_cache_dir is NULL terminated
+    tmp_cache_dir[len - 1] = '\0';
```
Comment (dirk-thomas) [inline src/rospack.cpp]: Since `strncpy` is already invoked with `size() + 1` this seems redundant.

File: src/rospack.cpp [hunk 4]
```diff
@@ -1994,23 +1994,33 @@ Rosstackage::writeCache()
...
     snprintf(tmp_cache_path, sizeof(tmp_cache_path), "%s\\.rospack_cache.XXXXXX", full_dir);
 #elif defined(__MINGW32__)
+    char tmp_cache_path[PATH_MAX];
     char* temp_name = tempnam(dirname(tmp_cache_dir),".rospack_cache.");
     snprintf(tmp_cache_path, sizeof(tmp_cache_path), temp_name);
     delete temp_name;
 #else
-    snprintf(tmp_cache_path, sizeof(tmp_cache_path), "%s/.rospack_cache.XXXXXX", dirname(tmp_cache_dir));
+    char *tmp_cache_path = NULL;
+    char *temp_dirname = strdup(dirname(tmp_cache_dir));
```
Comment (dirk-thomas) [inline src/rospack.cpp]: What is the purpose of `temp_dirname`? Why not use `tmp_cache_dir` which it is copied from?

File: src/rospack.cpp [hunk 5]
```diff
@@ -2068,6 +2078,9 @@ Rosstackage::writeCache()
         }
       }
     }
+#if !defined(_MSC_VER) && !defined(__MINGW32__) && !defined(WIN32)
```
Comment (dirk-thomas) [inline src/rospack.cpp]: The variable is being allocated in the `else` block above which follows the following two conditions:
- `defined(_MSC_VER)`
- `defined(__MINGW32__)`

Why is this condition additionally checking for `!defined(WIN32)`?

File: src/rospack.cpp [hunk 6]
```diff
@@ -2068,6 +2078,9 @@ Rosstackage::writeCache()
...
+		 free(tmp_cache_path);
```
Comment (dirk-thomas) [inline src/rospack.cpp]: The indentation is off - should use spaces instead of tabs.

File: src/rospack.cpp (hunk 2 above)
Comment (dirk-thomas) [inline src/rospack.cpp]: I guess this should check for successful allocation.

**Comment by jspricke** [comment]
Thanks for the comments, I've fixed them and will squash the commits, if you are fine with the patch.

**Comment by dirk-thomas** [comment]
Thank you for iterating on this. With the fancy new feature of GitHub I could also do the squashing during the merge?

**Comment by jspricke** [comment]
Not sure about the commit message then, but sure go ahead :)

**Comment by dirk-thomas** [comment]
You can freely edit the message and description on merge.

**Comment by jspricke** [comment]
Great, thanks :)

**Pull request merged**